controllers/cloud.redhat.com/jobinvocation_controller.go

```diff
@@ -0,0 +1,240 @@
+	qualifiedName := fmt.Sprintf("%s:%s", req.Namespace, req.Name)
+	log := r.Log.WithValues("jobinvocation", qualifiedName)
+	ctx := context.WithValue(context.Background(), errors.ClowdKey("log"), &log)
+	ctx = context.WithValue(ctx, errors.ClowdKey("recorder"), &r.Recorder)
+	//proxyClient := ProxyClient{Ctx: ctx, Client: r.Client}
+	jinv := crd.JobInvocation{}
+	err := r.Client.Get(ctx, req.NamespacedName, &jinv)
+
+	if err != nil {
+		if k8serr.IsNotFound(err) {
+			// Must have been deleted
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	r.Log.Info("Reconciliation started", "JobInvocation", fmt.Sprintf("%s:%s", jinv.Namespace, jinv.Name))
+	ctx = context.WithValue(ctx, errors.ClowdKey("obj"), &jinv)
+
+	// Get the ClowdApp
+	app := crd.ClowdApp{}
+	err = r.Client.Get(ctx, types.NamespacedName{
+		Name:      jinv.Spec.AppName,
+		Namespace: req.Namespace,
+	}, &app)
+
+	// Determine if the ClowdApp containing the Job exists
+	if err != nil {
+		r.Recorder.Eventf(&jinv, "Warning", "ClowdAppMissing", "ClowdApp [%s] is missing; Job cannot be jinved", jinv.Spec.AppName)
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	// Determine if the Clowapp containing the Job is ready
+	if app.Status.Ready == false {
+		r.Recorder.Eventf(&app, "Warning", "ClowdAppNotReady", "ClowdApp [%s] is not ready", jinv.Spec.AppName)
+		r.Log.Info("App not yet ready, requeue", "jobinvocation", jinv.Spec.AppName, "namespace", app.Namespace)
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	// Get the ClowdEnv for InvokeJob
+	env := crd.ClowdEnvironment{}
+	err = r.Client.Get(ctx, types.NamespacedName{
+		Name: app.Spec.EnvName,
+	}, &env)
+
+	// Walk the job names to be invoked and match in the ClowdApp Spec
+	for _, job := range jinv.Spec.Jobs {
+		jobContent, err := matchAndReturnJob(job, &app)
+		if err != nil {
+			r.Recorder.Eventf(&app, "Error", "JobNameMissing", "ClowdApp [%s] has no job named", jinv.Spec.AppName, job)
+			r.Log.Info("Missing Job Definition", "jobinvocation", jinv.Spec.AppName, "namespace", app.Namespace)
+			return ctrl.Result{}, err
+		}
+		r.Log.Info("Invoking job", "jobinvocation", job, "namespace", app.Namespace)
+		if err := r.InvokeJob(jobContent, &app, &env, ctx); err != nil {
+			return ctrl.Result{Requeue: true}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *JobInvocationReconciler) InvokeJob(job crd.Job, app *crd.ClowdApp, env *crd.ClowdEnvironment, ctx context.Context) error {
+	now := time.Now()
+	nn := types.NamespacedName{
+		Name:      fmt.Sprintf("%v-%v-%v", app.Name, job.Name, now.Unix()),
+		Namespace: app.Namespace,
+	}
+	if !job.OnRequest {
+		r.Recorder.Eventf(app, "Warning", "Only OnRequest Jobs are supported at this time", "%s is not set to onRequest", job)
+		r.Log.Info("Unsupported Job Type", "jobinvocation", job.Name, "namespace", app.Namespace)
+	} else {
+		j := batchv1.Job{}
+		createJob(app, env, nn, job, &j)
```

Comment:
If this is successful, perhaps we should emit an event that the job was successfully invoked?